Test/comp/ComprehensionsNewSyntax.dfy

```diff
@@ -0,0 +1,284 @@
+
+// SetComprehension1 is like SetComprehension0, but also adds "null" to "s".
+method SetComprehension1() {
+  var w, x, y, z := new ClassA, new ClassA, new ClassB, new ClassB;
+  var s := {w, x, y, z, null};
+  // The following set comprehension picks att elements in s:
+  var all := set o: object <- s;
+  // The next set comprehension picks out 2 of the elements in s:
+  var aa := set o: ClassA <- s;
+  // The next set comprehension picks out the other 2 of the elements in s:
+  var bb := set o: ClassB <- s;
+  // The following picks out the same elements as in bb:
+  var nn := set o: NothingInParticular <- s;
+
+  print |s|, " ", |all|, " ";           // 5 4
+  print |aa|, " ", |bb|, " ";           // 2 2
+  print |aa * bb|, " ", |aa + bb|, " "; // 0 4
+  print |nn|, " ", bb == nn, "\n";      // 2 true
+}
+
+// SetComprehension2 is like SetComprehension1, but uses maybe-null types in comprehensions
+method SetComprehension2() {
+  var w, x, y, z := new ClassA, new ClassA, new ClassB, new ClassB;
+  var s := {w, x, y, z, null};
+  // The following set comprehension picks att elements in s:
+  var all := set o: object? <- s;
+  // The next set comprehension picks out 2 of the elements in s:
+  var aa := set o: ClassA? <- s;
+  // The next set comprehension picks out the other 2 of the elements in s:
+  var bb := set o: ClassB? <- s;
+  // The following picks out the same elements as in bb:
+  var nn := set o: NothingInParticular? <- s;
+
+  print |s|, " ", |all|, " ";           // 5 5
+  print |aa|, " ", |bb|, " ";           // 3 3
+  print |aa * bb|, " ", |aa + bb|, " "; // 1 5
+  print |nn|, " ", bb == nn, "\n";      // 3 true
+}
+
+datatype Color = Red | Green | Blue
+
+predicate method True<G>(g: G) { true }
+
+method SetComprehension3() {
+  var s: set<bool> := {false, true};
+  // The following set comprehension picks att elements in s:
+  var all := set o: bool <- s;
+  var aa := set o: bool <- s | !o;
+  var bb := set o: bool <- s | o;
+
+  print |s|, " ", |all|, " ";           // 2 2
+  print |aa|, " ", |bb|, " ";           // 1 1
+  print |aa * bb|, " ", |aa + bb|, " "; // 0 2
+  print aa == all, " ", aa <= all, "\n"; // false true
+
+  var d := set z: Color | True(z);
+  var e := set z: Color | z in d;
+  print |d|, " ", |e|, "\n"; // 3 3
+}
+
+trait ICell { var data: int }
+class CellA extends ICell { }
+class CellB extends ICell { }
+
+method Enumerations() {
+  var c, d, e := new CellA, new CellA, new CellB;
+  c.data, d.data, e.data := 4, 5, 1;
+  var s: set<ICell?> := {c, d, e, null};
+  print c.data, d.data, e.data, "\n";  // 451
+
+  // non-sequentialized forall statement
+  forall a: CellA <- s {
+    a.data := c.data + a.data - 2;
+  }
+  print c.data, d.data, e.data, "\n";  // 671
+
+  // sequentialized forall statement
+  forall a: CellA <- s {
+    a.data := 2;
+  }
+  print c.data, d.data, e.data, "\n";  // 221
+
+  // assign-such-that statement
+  d.data := 9;
+  assert d in s;
+  var u: CellA :| u in s && 7 <= u.data;
```

Comment:
Could you have:
```suggestion
  var u: CellA <- s | 7 <= u.data;
```
?

Reply:
I had exactly the same thought. :) Or perhaps the syntax `var u: CellA <- s :| 7 <= u.data;`

But given this wasn't in scope in the RFC, and let/assign-such-that isn't already using the `QuantifierDomain` syntax, I don't want to pull that in to this step just yet. But it's really tempting as a follow-up, especially because it then allows deterministic behavior (select the first matching value in `s`) without having to prove there exists exactly one such value.

Reply:
I have to admit the more I think about this the more I like it. Essentially `var` would be interpreted as another "binder" option like `forall` or `set`, and in this case it means "take the first value(s) such that...". I even like the way just `var x <- s;` reads as "let x be some element of s".

Reply:
I did not think about the "first" being interpreted as a "uniqueness", but it makes even more sense. I love it.